wurst/systems/entities/bosses/TheOne.wurst

```diff
@@ -82,40 +81,19 @@ public function spawnTheOne()
 public function canSpawnTheOne() returns boolean
     return theOne == null or not theOne.getUnit().isAlive()
 
-function checkInvulnerability() returns bool
-    // The One is always invulnerable when outside its arena.
-    if not theOne.getUnit().isInRegion(THE_ONE_REGION)
-        return true
-
-    // The One is invulnerable unless there is a troll present in its arena.
-    for troll in getTrolls()
-        // Disable invulnerability if both conditions are met.
-        if troll.isAlive() and troll.isInRegion(THE_ONE_REGION)
-            return false
-
-    // Enable invulnerability if no matches were found.
-    return true
-
-function toggleInvulnerability()
-    // Exit if The One is not currently alive.
-    if theOne == null or not theOne.getUnit().isAlive()
-        return
-
-    // Toggle the invulnerability.
-    if checkInvulnerability()
-        theOne.getUnit().addAbility(AbilityIds.invulnerable)
-    else
-        theOne.getUnit().removeAbility(AbilityIds.invulnerable)
-
 init
     TheOne.initialize()
     CreateTrigger()
         ..registerEnterRegion(THE_ONE_REGION, null)
-        ..addAction(function toggleInvulnerability)
+        ..addAction()->
+        if theOne != null
+            toggleInvulnerability(THE_ONE_REGION, theOne.getUnit())
     CreateTrigger()
         ..registerLeaveRegion(THE_ONE_REGION, null)
-        ..addAction(function toggleInvulnerability)
+        ..addAction()->
+        if theOne != null
+            toggleInvulnerability(THE_ONE_REGION, theOne.getUnit())
     registerPlayerUnitEvent(
-        EVENT_PLAYER_UNIT_DEATH,
-        function toggleInvulnerability
-    )
+        EVENT_PLAYER_UNIT_DEATH)->
+        if theOne != null
```

Comment:
I would leave the check for null in `toggleInvulnerability`. It's cheap and simplifies the logic for conditional bosses.

Reply:
right now toggleInvulnerability is given a unit and region, doesnt it make sense to keep that check outside the function since not all bosses we will be checking have an object like that?

Reply:
It doesn't complicate the logic and is simpler to write once. Plus it works as a safeguard against future mistakes where you might forget to check for null otherwise.

Reply:
mmk, what would i use to accept any class? ie. TheOne or DiscoDuck

Reply:
It would still take a `unit`. I just mean that you should add the null check there. `unit` can also be null.

Reply:
Ah, I see what the problem is. You would still need the null check on `theOne` before calling `getUnit`. I would add a second function that takes `UnitEntity` and just forwards the call to the function you have now if the entity isn't null.